pkg/providers/vault/provider.go

```diff
@@ -0,0 +1,370 @@
+			resp, b.String())
+	}
+
+	var d struct {
+		Data struct {
+			Data map[string]string `json:"data"`
+		} `json:"data"`
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&d); err != nil {
+		return "", errors.Wrapf(err, "failed to read body")
+	}
+
+	return d.Data.Data[secretName], nil
+}
+
+func (p *Provider) getRootCAsPools() (*x509.CertPool, error) {
+	switch {
+	case p.VaultCAPem != "":
+		certPool := x509.NewCertPool()
+		if err := loadCert(certPool, []byte(p.VaultCAPem)); err != nil {
+			return nil, err
+		}
+		return certPool, nil
+	case p.VaultCAPath != "":
+		certPool := x509.NewCertPool()
+		if err := loadCertFolder(certPool, p.VaultCAPath); err != nil {
+			return nil, err
+		}
+		return certPool, nil
+	case p.VaultCACert != "":
+		certPool := x509.NewCertPool()
+		if err := loadCertFile(certPool, p.VaultCACert); err != nil {
+			return nil, err
+		}
+		return certPool, nil
+	default:
+		certPool, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't load system certs")
+		}
+		return certPool, err
+	}
+}
+
+// loadCert loads a single pem-encoded certificate into the given pool.
+func loadCert(pool *x509.CertPool, pem []byte) error {
+	if ok := pool.AppendCertsFromPEM(pem); !ok {
+		return fmt.Errorf("failed to parse PEM")
+	}
+	return nil
+}
+
+// loadCertFile loads the certificate at the given path into the given pool.
+func loadCertFile(pool *x509.CertPool, p string) error {
+	pem, err := ioutil.ReadFile(p)
+	if err != nil {
+		return errors.Wrapf(err, "couldn't read CA file from disk")
+	}
+
+	if err := loadCert(pool, pem); err != nil {
+		return errors.Wrapf(err, "couldn't load CA at %s", p)
+	}
+
+	return nil
+}
+
+// loadCertFolder iterates exactly one level below the given directory path and
+// loads all certificates in that path. It does not recurse
+func loadCertFolder(pool *x509.CertPool, p string) error {
+	if err := filepath.Walk(p, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		return loadCertFile(pool, path)
+	}); err != nil {
+		return errors.Wrapf(err, "failed to load CAs at %s", p)
+	}
+	return nil
+}
+
+// MountSecretsStoreObjectContent mounts content of the vault object to target path
+func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib map[string]string, secrets map[string]string, targetPath string, permission os.FileMode) (err error) {
+	roleName := attrib["roleName"]
+	if roleName != "" {
+		p.VaultRole = roleName
+	}
+	glog.V(2).Infof("vault: roleName %s", p.VaultRole)
+
+	p.VaultAddress = attrib["vaultAddress"]
+	if p.VaultAddress == "" {
+		p.VaultAddress = defaultVaultAddress
+	}
+	glog.V(2).Infof("vault: vault address %s", p.VaultAddress)
+
+	p.VaultCAPem = attrib["vaultCAPem"]
+	p.VaultCACert = attrib["vaultCACertPath"]
+	p.VaultCAPath = attrib["vaultCADirectory"]
+	p.VaultServerName = attrib["vaultTLSServerName"]
+
```

Comment:
Do we need to ensure these parameters are not empty if `VaultSkipVerify == false`

Reply:
We need one of the values to be set. Otherwise, we default to system certificates. https://github.com/deislabs/secrets-store-csi-driver/pull/23/files/aec594d0f2d07108eff7186238ed117dd43e6b02#diff-e516a1abcc9e4dff2b12ce96f4720d28R204